# Asg-10 focuses on questions graded by the mentioned TAs

 

In [41]:
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
import sys

# -- Detect if in Own Install or in Colab
try:
    import google.colab
    OWN_INSTALL = False
except:
    OWN_INSTALL = True
    
if OWN_INSTALL:
    print("Own install")
    sys.path[0:0] = ['../../../../..',  '../../../../../3rdparty',
                     '../../../..',  '../../../../3rdparty',
                     '../../..',     '../../../3rdparty',
                     '../..',        '../../3rdparty',
                     '..',           '../3rdparty' ]
    
else: # In colab
    ! if [ ! -d Jove ]; then git clone https://github.com/ganeshutah/Jove Jove; fi
    sys.path.append('./Jove')
    sys.path.append('./Jove/jove')

# -- common imports --
from jove.PcpJupyterNew import *
#~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Own install


# =============

In [42]:
help(pcp_solve)

Help on function pcp_solve in module jove.PcpJupyterNew:

pcp_solve(pcp_pairs, OWN_INSTALL=False, run=None, ni=False, di=None, depth=None, tiles_per_row=15)
    Forward user input to a file, which we then use Ling Zhao's pcp solver to solve.
    :param OWN_INSTALL: if True, running on one's own machine; default is Colab (so false)
    :param pcp_pairs: List of tuple pairs representing pcp 'tiles'
    :param run: Number of runs to perform.
    :param ni: No iterative search.
    :param di: Depth increment.
    :param depth: Search depth.
    :param tiles_per_row: Number of tiles to show in single row together (useful for creating more meaningful output)



# (1) LT, 25%:
 

## 1(a): 

In [43]:
# Tool Solution
pcp_solve([('110', '1'), ('1', '0'), ('0', '110')], OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[3, 2, 2, 3, 3, 1, 3, 2, 2, 2, 1, 3, 3, 3, 2, 1, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 1, 3, 2, 3, 2, 2, 1, 3, 2, 1, 3, 3, 2, 2, 3, 3, 2, 1, 3, 2, 1, 1, 3, 3, 1, 1, 3, 2, 1, 1, 1, 1, 2, 2, 1, 1, 2, 2, 3, 1, 3, 3, 1, 2, 1, 1, 2, 1, 1]
0   1   1   0   0   011 0   1   1   1   011 0   0   0   1   
011 0   0   011 011 1   011 0   0   0   1   011 011 011 0   

011 0   1   1   011 0   1   011 0   0   1   011 0   1   0   
1   011 0   0   1   011 0   1   011 011 0   1   011 0   011 

1   1   011 0   1   011 0   0   1   1   0   0   1   011 0   
0   0   1   011 0   1   011 011 0   0   011 011 0   1   011 

1   011 011 0   0   011 011 0   1   011 011 011 011 1   1   
0   1   1   011 011 1   1   011 0   1   1   1   1   0   0   

011 011 

Solution length: 75 for both solutions

## 1(b): 

In [44]:
# Tool solution
pcp_solve([('110', '1'), ('1', '0'), ('0', '110'), ('0', '1')],
          OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[3, 2, 2, 4, 1]
0   1   1   0   011 
011 0   0   1   1   




Solution length: 5

## 1(c, hand): 

<!-- Solution by hand is 2, 1, 3 -->

We have the following tiles:

  * Tile 1: `(1, 0)`

  * Tile 2: `(101, 1)`

  * Tile 3: `(0, 110)`

Solution construction:

  * Tile 2 is the only tile where both the top and the bottom start with the same character ("1"), therefore it must go first:
    * Solution so far: `(101, 1)`

  * The top now has an extra zero following, so tile 1 must go next since it's the only one having the bottom starting with a zero.
    * Solution so far: `(101, 1) (1, 0)`

  * The top now has an two extra ones following, so it's best for tile 3 to go next since it starts with two ones on the bottom.
    * Solution so far: `(101, 1) (1, 0) (0, 110)`

  * Now, the top and the bottom match up. The final solution is `(101, 1) (1, 0) (0, 110)` or tiles 2, 1, 3.

## 1(c, tool check): 

In [45]:
# Solution by tool for the same
pcp_solve([('1', '0'), ('101', '1'), ('0', '110')], OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[3, 1, 2]
0   1   101 
011 0   1   




The tool presented the solution in reverse, but we can see that it's the same solution.

## 1(d): 

In [57]:
# Create the tiles separately; easier to change and run that way
# Found it hard to create own instances that explode

TILES = [('101010101010', '10'), ('010101010101', '10'), ('01', '0101')]
pcp_solve(TILES, OWN_INSTALL)

Detected platform windows; PLEASE be running with Admin Privileges!
 Running the command ... :  ['pcp.exe', '-i', 'temp.txt']
Solution(s) to PCP instance are below. Note: the tiles may be reversed,
  as the solver may sometimes present the solution in reverse.:

Solution 1
[1, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3]
101010101010 010101010101 010101010101 010101010101 010101010101 010101010101 01           01           01           01           01           01           01           01           01           
10           10           10           10           10           10           0101         0101         0101         0101         0101         0101         0101         0101         0101         

01           01           01           01           01           01           01           01           01           01           01           01           01           01           01           
0101         0101         01

Solution length: 36 for all six solutions

# =============

# (2a) AR, 15%:

* Suppose there exists a decider $H$ for $\textit{Halt}_\text{TM}$.

  * Since $H$ is a decider, $H$ will always halt, and $H$ will accept $\left<M, w\right>$ if and only if $\left<M, w\right> \in \textit{Halt}_\text{TM}$, or in other words, $M$ halts on $w$.

* Create a program $D$ that takes input of a Turing machine $T$. Let $D$ halt and accept if $H(\left<T, T\right>)$ rejects, and let $D$ loop otherwise.

  * This means that $D$ halts if the Turing machine $T$ loops on input $T$ and loops if $T$ halts on $T$.

* Run the program $D$ with input of the $D$ itself. There are two possibilities of what $D$ would do:

  * $D$ halts and accepts, meaning that $H(\left<D, D\right>)$ rejected, but this rejection would mean that $D$ should not halt on input $D$, which creates a contradiction.

  * $D$ loops, meaning that $H(\left<D, D\right>)$ accepted, but this acceptance would mean that $D$ should halt and accept on input $D$, which creates a contradiction.

* Therefore, we obtain a contradiction under all cases.
  
* Thus, $\textit{Halt}_\text{TM}$ is undecidable. $\square$


# (2b) AR, 10%:

* The mapping reduction from $A_\text{TM}$ to $\textit{CFL}_\text{TM}$ is to emit this machine $M'$:

  * $M'$ takes input of a string $x$.

  * If $x$ is in the form of $0^n1^n2^n$, then $M'$ rejects $x$ immediately.

  * Otherwise, $M'$ runs $M$ on $w$:

    * If $M$ accepts $w$, then $M'$ accepts $x$.

    * If $M$ rejects $w$, then $M'$ rejects $x$.

* Assume that there exists a decider $D$ that decides whether $M'$ would accept some input $x$ or not:

  * If $D$ accepts, then $M'$ has a CFL, and $M$ must've accepted $w$.

  * If $D$ rejects, then $M'$ does not have a CFL, and $M$ must've rejected $w$.
  
* Thus, we have achieved our mapping reduction.

  * That is, $M'$ has a CFL if and only if $M$ accepts $w$.

# =============

# (3, XL, SV) 50%:  


## 3.a.i - describe $A, B, f$

* What is A in this problem ? Describe clearly.

* What is B in this problem ? Describe clearly.

* What is f in this problem ? Describe clearly.

## 3.a.ii - Proof for the $x \in A  \Rightarrow  f(x) \in B$ part

* Proof steps - be thorough; it is worth 25 points!

* 

*
 

## 3.b
 
* Proof steps - be thorough; it is worth 25 points!

*

*



# END
